File: PartB/ask2/main.cpp
#include <iostream>
#include <string.h>
#include <fstream>
#include <sstream>
#include <vector>

using namespace std;

struct Data {
    string Direction;
    int Year;
    string Date;
    string Weekday;
    string Country;
    string Commodity;
    string Transport_Mode;
    string Measure;
    long int Value;
    long int Cumulative;
};

struct Node {
    Data data;
    Node *left;
    Node *right;
    int height;
};

// Calculate height
int height(Node *node) {
    if (node == NULL)
        return 0;
    return node->height;
}

int max(int a, int b) {
    return (a > b) ? a : b;
}

// New node creation
Node *newNode(Data data) {
    Node *node = new Node();
    node->data = data;
    node->left = NULL;
    node->right = NULL;
    node->height = 1;
    return (node);
}

// Rotate right
Node *rightRotate(Node *y) {
    Node *x = y->left;
    Node *T2 = x->right;

    x->right = y;
    y->left = T2;

    y->height = max(height(y->left), height(y->right)) + 1;
    x->height = max(height(x->left), height(x->right)) + 1;

    return x;
}

// Rotate left
Node *leftRotate(Node *x) {
    Node *y = x->right;
    Node *T2 = y->left;

    y->left = x;
    x->right = T2;

    x->height = max(height(x->left), height(x->right)) + 1;
    y->height = max(height(y->left), height(y->right)) + 1;

    return y;
}

// Get balance factor of each node
int getBalance(Node *node) {
    if (node == NULL)
        return 0;
    return height(node->left) - height(node->right);
}

// Insert node
Node *insertNode(Node *node, Data data) {
    if (node == NULL)
        return (newNode(data));
    if (data.Value < node->data.Value)
        node->left = insertNode(node->left, data);
    else if (data.Value > node->data.Value)
        node->right = insertNode(node->right, data);
    else
        return node;

    // Update the balance factor of each node and balance the tree
    node->height = 1 + max(height(node->left), height(node->right));
    int balance = getBalance(node);
    if (balance > 1) {
        if (data.Value < node->left->data.Value) {
            return rightRotate(node);
        } else if (data.Value > node->left->data.Value) {
            node->left = leftRotate(node->left);
            return rightRotate(node);
        }
    }
    if (balance < -1) {
        if (data.Value > node->right->data.Value) {
            return leftRotate(node);
        } else if (data.Value < node->right->data.Value) {
            node->right = rightRotate(node->right);
            return leftRotate(node);
        }
    }
    return node;
}

// Node with minimum value
Node *nodeWithMinValue(Node *node) {
    Node *current = node;
    while (current->left != NULL)
        current = current->left;
    return current;
}

Node *nodeWithMaxValue(Node *node) {
    Node *current = node;
    while (current->right != NULL)
        current = current->right;
    return current;
}

Node *read_data(string filename) {
    Node *root = NULL;
    ifstream file(filename);
    if (!file.is_open()) {
        cout << "Error opening file" << endl;
        exit(1);
    }
    string line;
    getline(file, line);
    while (getline(file, line)) {
        Data temp;
        string temp_array[10];

        stringstream ss(line);
        string token;

        bool insideQuotes = false;
        int i = 0;
        while (getline(ss, token, ',')) {
            if (token.front() == '\"' && !insideQuotes) {
                insideQuotes = true;
            }
        
            if (token.back() == '\"' && insideQuotes) {
                insideQuotes = false;
            }
        
            if (insideQuotes) {
                temp_array[i - 1] += "," + token;
            } else {
                temp_array[i] = token;
                i++;
            }
        }

        temp.Direction = temp_array[0];
        temp.Year = stoi(temp_array[1]);
        temp.Date = temp_array[2];
        temp.Weekday = temp_array[3];
        temp.Country = temp_array[4];
        temp.Commodity = temp_array[5];
        temp.Transport_Mode = temp_array[6];
        temp.Measure = temp_array[7];
        temp.Value = stol(temp_array[8]);
        temp.Cumulative = stol(temp_array[9]);

        root = insertNode(root, temp);
        
    }
    file.close();
    return root;
}

void preorder(Node *root) {
    if (root == NULL) {
        return;
    }
    cout << root->data.Direction << ", "
            << root->data.Year << ", "
            << root->data.Date << ", "
            << root->data.Weekday << ", "
            << root->data.Country << ", "
            << root->data.Commodity << ", "
            << root->data.Transport_Mode << ", "
            << root->data.Measure << ", "
            << root->data.Value << ", "
            << root->data.Cumulative << endl;
    preorder(root->left);
    preorder(root->right);
}

void printDates(Node *root, string date) {
    if (root == NULL) {
        return;
    }
    if (root->data.Date == date) {
        cout << root->data.Direction << ", "
            << root->data.Year << ", "
            << root->data.Date << ", "
            << root->data.Weekday << ", "
            << root->data.Country << ", "
            << root->data.Commodity << ", "
            << root->data.Transport_Mode << ", "
            << root->data.Measure << ", "
            << root->data.Value << ", "
            << root->data.Cumulative << endl;
    }
    printDates(root->left, date);
    printDates(root->right, date);
}

int getLength(Node *root) {
    if (root == NULL) {
        return 0;
    }
    return 1 + getLength(root->left) + getLength(root->right);
}

int main() {
    Node *root = read_data("effects.csv");

    cout << "Number of nodes: " << getLength(root) << endl;
    
    while (true) {
        cout << "\n--------------------------------" << endl;
        cout << " MENU " << endl;
        cout << "--------------------------------" << endl;
        cout << "1. find the date with the Max value " << endl;
        cout << "2. find the date with the Min value " << endl;
        cout << "3. exit " << endl;
        cout << "enter your choice: ";
        int choice;
        cin >> choice;
        switch (choice) {
            case 1: {
                Node *maxNode = nodeWithMaxValue(root);
                cout << "\n\tDate with Max value: " << maxNode->data.Date << endl;
                printDates(root, maxNode->data.Date);
                break;
            }
            case 2: {
                Node *minNode = nodeWithMinValue(root);
                cout << "\n\tDate with Min value: " << minNode->data.Date << endl;
                printDates(root, minNode->data.Date);
                break;
            }
            case 3: {
                exit(0);
            }
            default: {
                cout << "\n\tInvalid choice" << endl;
                break;
            }
        }
    }
    
    return 0;
}

File: PartB/ask2/src/main.rs
#[allow(unused)]
use std::fs::File;
use std::io::Write;
use std::process::exit;
use std::time::SystemTime;
use std::rc::Rc;

mod tests;

#[derive(Debug, Clone)]
struct Data {
    direction: String,
    year: u16,
    date: String,
    weekday: String,
    country: String,
    comodity: String,
    transport_mode: String,
    measure: String,
    value: u64,
    cumulative: u64,
}

#[derive(Debug, Clone)]
struct Node {
    data: Rc<Data>,
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
    height: i32,
}

#[derive(Debug)]
struct AvlTree {
    root: Option<Box<Node>>,
}

impl AvlTree {
    fn new() -> AvlTree {
        AvlTree { root: None }
    }

    fn insert(&mut self, data: Rc<Data>) {
        self.root = insert(&mut self.root, data);
    }
}

fn height(node: &Option<Box<Node>>) -> i32 {
    match node {
        Some(n) => n.height,
        None => 0,
    }
}

fn balance_factor(node: &Node) -> i32 {
    height(&node.left) - height(&node.right)
}

fn update_height(node: &mut Box<Node>) {
    let hl = height(&node.left);
    let hr = height(&node.right);
    node.height = std::cmp::max(hl, hr) + 1;
}

fn rotate_left(mut node: Box<Node>) -> Box<Node> {
    let mut new_root = node.right.take().unwrap();
    node.right = new_root.left.take();
    update_height(&mut node);
    update_height(&mut new_root);
    new_root.left = Some(node);
    new_root
}

fn rotate_right(mut node: Box<Node>) -> Box<Node> {
    let mut new_root = node.left.take().unwrap();
    node.left = new_root.right.take();
    update_height(&mut node);
    update_height(&mut new_root);
    new_root.right = Some(node);
    new_root
}

fn balance(mut node: Box<Node>) -> Box<Node> {
    if balance_factor(&node) > 1 {
        if balance_factor(&node.left.as_ref().unwrap()) < 0 {
            node.left = Some(rotate_left(node.left.unwrap()));
        }
        return rotate_right(node);
    }
    if balance_factor(&node) < -1 {
        if balance_factor(&node.right.as_ref().unwrap()) > 0 {
            node.right = Some(rotate_right(node.right.unwrap()));
        }
        return rotate_left(node);
    }
    update_height(&mut node); // Moved here to avoid redundant height updates
    node
}

fn insert(root: &mut Option<Box<Node>>, data: Rc<Data>) -> Option<Box<Node>> {
    if root.is_none() {
        return Some(Box::new(Node {
            data: data, // No need to clone
            left: None,
            right: None,
            height: 0,
        }));
    }
    if data.value < root.as_ref().unwrap().data.value {
        root.as_mut().unwrap().left = insert(&mut root.as_mut().unwrap().left, data.clone());
    } else {
        root.as_mut().unwrap().right = insert(&mut root.as_mut().unwrap().right, data.clone());
    }
    Some(balance(root.take().unwrap()))
}

fn node_with_min_value(root: &Option<Box<Node>>) -> Option<Box<Node>> {
    let mut current = root.as_ref().unwrap();
    while let Some(node) = &current.left {
        current = node;
    }
    return Some(current.clone());
}

fn node_with_max_value(root: &Option<Box<Node>>) -> Option<Box<Node>> {
    let mut current = root.as_ref().unwrap();
    while let Some(node) = &current.right {
        current = node;
    }
    return Some(current.clone());
}

fn nodes_with_same_value<'a>(root: &'a Option<Box<Node>>, value: &u64, nodes: &mut Vec<&'a Node>) {
    if let Some(ref node) = *root {
        if &node.data.value == value {
            nodes.push(node.as_ref());
        }
        nodes_with_same_value(&node.left, value, nodes);
        nodes_with_same_value(&node.right, value, nodes);
    }
}

fn read_data(filename: &str) -> Option<AvlTree> {
    let mut reader = match csv::Reader::from_path(filename) {
        Ok(reader) => reader,
        Err(_) => {
            println!("Error reading file");
            exit(1);
        }
    };
    let mut tree = AvlTree::new();

    for result in reader.records() {
        let record = match result {
            Ok(record) => record,
            Err(_) => {
                println!("Error reading record");
                exit(1);
            }
        };
        let data = Rc::new(Data {
            direction: record.get(0).unwrap().to_string(),
            year: record.get(1).unwrap().parse::<u16>().unwrap(),
            date: record.get(2).unwrap().to_string(),
            weekday: record.get(3).unwrap().to_string(),
            country: record.get(4).unwrap().to_string(),
            comodity: record.get(5).unwrap().to_string(),
            transport_mode: record.get(6).unwrap().to_string(),
            measure: record.get(7).unwrap().to_string(),
            value: record.get(8).unwrap().parse::<u64>().unwrap(),
            cumulative: record.get(9).unwrap().parse::<u64>().unwrap(),
        });

        tree.insert(data);
    }

    return Some(tree);
}

fn user_input() -> String {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    input = input.trim().to_string();
    input
}

fn print_data(data: &Data) {
    println!(
        "{}, {}, {}, {}, {}, {}, {}, {}, {}, {}",
        data.direction,
        data.year,
        data.date,
        data.weekday,
        data.country,
        data.comodity,
        data.transport_mode,
        data.measure,
        data.value,
        data.cumulative
    );
}

fn main() {
    let start = SystemTime::now();
    let root = read_data("effects.csv");
    println!("Time elapsed: {:?}", start.elapsed().unwrap());
    
    loop {
        println!("---------------------------");
        println!("1. find the data with the MAX value");
        println!("2. find the data with the MIN value");
        println!("0. exit");
        println!("---------------------------");
        print!("Enter your choice: ");
        std::io::stdout().flush().unwrap();
        let choice = user_input();

        match choice.as_str() {
            "1" => {
                let node = node_with_max_value(&root.as_ref().unwrap().root);
                match node {
                    Some(node) => {
                        let mut nodes = Vec::new();
                        nodes_with_same_value(&root.as_ref().unwrap().root, &node.data.value, &mut nodes);
                        nodes.into_iter().take(10).for_each(|node| {
                            print_data(&node.data);
                        });
                    }
                    None => {
                        println!("No data found");
                    }
                }
            }
            "2" => {
                let node = node_with_min_value(&root.as_ref().unwrap().root);
                match node {
                    Some(node) => {
                        let mut nodes = Vec::new();
                        nodes_with_same_value(&root.as_ref().unwrap().root, &node.data.value, &mut nodes);
                        nodes.into_iter().take(10).for_each(|node| {
                            print_data(&node.data);
                        });
                    }
                    None => {
                        println!("No data found");
                    }
                }
            }
            "0" => {
                break;
            }
            _ => {
                println!("Invalid choice");
            }
        }
        println!()
    }

    println!("Bye!");
}
